stdlib/public/core/StringWordView.swift

```diff
@@ -0,0 +1,202 @@
+      _internalInvariant($0 >= 0)
+
+      guard $0 < _guts.count else {
+        return nil
+      }
+
+      let scalars = String.UnicodeScalarView(_guts)
+      let idx = String.Index(_encodedOffset: $0)
+
+      let scalar = scalars[idx]
+      let nextIndex = scalars.index(after: idx)
+
+      return (scalar, nextIndex._encodedOffset)
+    }
+
+    let nextIndex = String.Index(_encodedOffset: nextOffset)._scalarAligned
+    return _guts.markEncoding(nextIndex)
+#else
+    fatalError("No foreign strings on this platform in this version of Swift.")
+#endif
+  }
+
+  @_spi(_Unicode)
+  @available(SwiftStdlib 5.7, *)
+  public subscript(position: Index) -> Element {
+    let position = _guts.validateWordIndex(position)
+    let indexAfter = _uncheckedIndex(after: position)
+
+    return String(_guts)[position ..< indexAfter]
+  }
+}
+
+@available(SwiftStdlib 5.7, *)
+extension String._WordView: BidirectionalCollection {
+  @_spi(_Unicode)
+  @available(SwiftStdlib 5.7, *)
+  public func index(before i: Index) -> Index {
+    let i = _guts.validateInclusiveWordIndex(i)
+    return _uncheckedIndex(before: i)
+  }
+
+  internal func _uncheckedIndex(before i: Index) -> Index {
+    _internalInvariant(_guts.hasMatchingEncoding(i))
+    _internalInvariant(i > startIndex)
+    _internalInvariant(i._isScalarAligned)
+
+    if _slowPath(_guts.isForeign) {
+      return _foreignIndex(before: i)
+    }
+
+    let previousOffset = _guts.withFastUTF8 { utf8 in
+      previousBoundary(endingAt: i._encodedOffset) {
+        _internalInvariant($0 <= _guts.count)
+
+        guard $0 > 0 else {
+          return nil
+        }
+
+        let (scalar, len) = _decodeScalar(utf8, endingAt: $0)
+        return (scalar, $0 &- len)
+      }
+    }
+
+    let previousIndex = String.Index(
+      _encodedOffset: previousOffset
+    )._scalarAligned
+    return _guts.markEncoding(previousIndex)
+  }
+
+  @inline(never)
+  internal func _foreignIndex(before i: Index) -> Index {
+#if _runtime(_ObjC)
+    let previousOffset = previousBoundary(endingAt: i._encodedOffset) {
+      _internalInvariant($0 <= _guts.count)
+
+      guard $0 > 0 else {
+        return nil
+      }
+
+      let scalars = String.UnicodeScalarView(_guts)
+      let idx = String.Index(_encodedOffset: $0)
+
+      let previousIndex = scalars.index(before: idx)
+      let scalar = scalars[previousIndex]
+
+      return (scalar, previousIndex._encodedOffset)
+    }
+
+    let previousIndex = String.Index(
+      _encodedOffset: previousOffset
+    )._scalarAligned
+    return _guts.markEncoding(previousIndex)
+#else
+    fatalError("No foreign strings on this platform in this version of Swift.")
+#endif
+  }
+}
+
+extension String {
+  @_spi(_Unicode)
+  @available(SwiftStdlib 5.7, *)
+  public var _words: _WordView {
+    _WordView(_guts)
+  }
+}
+
+extension String {
+  @_spi(_Unicode)
+  @available(SwiftStdlib 5.7, *)
+  public func _isOnWordBoundary(_ i: String.Index) -> Bool {
```

Comment:
What's the intended use case for this? (We won't be calling it in a loop for every index in some string, right?)

Reply:
OK, so is `\b` going to call this on every index then?

I'm asking because rounding down operates by doing an index(before:) + index(after:) dance, which (when done repeatedly) is going to be significantly slower than memoizing where the nearest word boundaries are.

Reply:
> OK, so is \b going to call this on every index then?

I mean, it's going to do what the regex algorithm requests that it does. By itself, no it will only call it once at the current position resulting in success or failure. If it's the first candidate in an alternation like `/(?:\b|.)*/` then it will get called until it succeeds, which is what the algorithm is specifying.

Reply:
My point is that this entry point laboriously calculates the nearest boundary only to throw it away immediately after comparing it to the passed-in index.

Replacing this with an entry point that rounds an index down (or up) to the nearest word boundary would allow us to memoize the results in the regex library, and therefore avoid needless repetitions of all this work for successive indices within the same word. (We would be far better off if we could amortize the cost of `_guts.roundDownToNearestWord(i)` across multiple invocations, rather than repeating it.)

Of course, if this is only a temporary implementation, and we'll be able to replace it with an O(1) variant later, then this would still make a plausible entry point.